python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -235,3 +235,13 @@ cdef _get_np_scalar_from_timestamp64(unique_ptr[scalar]& s):
         )
     else:
         raise ValueError("Could not convert cudf::scalar to numpy scalar")
+
+
+def as_scalar(val, dtype=None):
+    if isinstance(val, Scalar):
+        if (dtype is None or dtype == val.dtype):
+            return val
+        else:
+            return Scalar(val.value, dtype)
+    else:
+        return Scalar(val, dtype)
```

Comment:
cc @shwina any opinions here on the scalar factory? It would be better if we could just bake this into a `Scalar.__new__` method from my perspective so we could do `Scalar(...)` instead of having to call `as_scalar` everywhere.

Reply:
Or could it go in `Scalar.__init__`?

Reply:
Can't go in `__init__` as we can't return reference.

Can't use `__new__` as Scalar is defined as `cdef` and the alternative `__cinit__` is a void function, so can't return.

Reply:
IIUC the issue is the case where the object is already as `Scalar`. Mainly this is an issue because we have a `unique_ptr` under the hood. Is that right?

What if we had a Cython class that held the `unique_ptr` as part of the internal spec (say `_ScalarWrapper` or similar) and made `Scalar` a Python or Cython class that just held a `_ScalarWrapper` instance. This way users can duplicate `Scalar` many times, but it always just points to the same `_ScalarWrapper` (as Python assignment is reference based). Would this work?

Reply:
Yes that would work. I'm leaning towards doing this as a follow up though as we want to get the performance improvements from this PR in ASAP.

Reply:
I will push a PR as a follow-up for this.

Reply:
Yep sounds reasonable.